design-proposals/klip-12-pull-high-availability.md

```diff
@@ -0,0 +1,290 @@
+  1) One of `B` or `C` is picked as the new active, this takes ~10 seconds based on default configs. 
+  2) Once a new active is chosen, depending on how far behind (lag) it is with respect to the changelog, 
+  it may take up  few seconds or even minutes before the standby has fully caught up to the last committed offset. 
+
+In total, it takes >>10 seconds before a query can start succeeding again. What can we do until a
+new active is ready to serve requests? 
+[KIP-535](https://cwiki.apache.org/confluence/display/KAFKA/KIP-535%3A+Allow+state+stores+to+serve+stale+reads+during+rebalance) 
+has laid the groundwork to allow us to serve requests from standbys, even if they are still in 
+rebalancing state or have not caught up to the last committed offset of the active. 
+
+Every ksql server now has the information of:
+
+1. Local offset lag: The lag of all partitions the server hosts (whether active or standby)
+2. Routing Metadata for a key: The partition, active host and standy hosts per key
+
+This information allows each server (with some communication, discussed below) to know the global 
+lag information of every topic partition whether it is the standby or active. This enables us to 
+implement a policy where `B` having established that `A` is down (after trying to send a request 
+to `A` and seeing it failed), to decide whether it (`B`) will serve it or `C`. This effectively 
+means that there will be little to no down time in serving requests at the cost of consistency as
+`B` or `C` may serve stale (not caught up) data. Hence, we achieve high availability at the cost of 
+consistency. Eventual consistency for the win!
+
+
+## What is in scope
+Ensure availability for pull queries when KSQL servers fail (with `ksql.streams.num.standby.replicas=N`, 
+we can tolerate upto N such failures) 
+
+## What is not in scope
+
+- Address unavailability caused by cold starts of KSQL server i.e when a new server is added to a 
+KSQL cluster, it must first rebuild its state off the changelog topics and that process still could 
+take a long time. 
+- Try to improve consistency guarantees provided by KSQL i.e reduce the amount of time it takes to 
+rebalance or standby replication to catch up. 
+
+## Value/Return
+
+The cluster of KSQl server will be able to serve requests even when the active is down. This mitigates 
+a large current gap in deploying KSQL pull queries for mission critical use-cases, by significantly 
+reducing failure rate of pull queries during server failures.
+
+## Public APIS
+No change to public APIs
+
+## Design
+
+The problem at hand is how to determine if a server is down and where to route a request when it is 
+down. Every KSQL server must have available the information of what other KSQL servers exist in the 
+cluster, their status and their lag per topic partition. There are three components to this: 
+Healthchecking, Lag reporting and Lag-aware routing.  
+
+### Failure detection/Healthcheck
```

Comment:
Not sure what the benefit is if we use a Kafka topic for heartbeats versus RPC calls since the mechanism of heartbeat processing is still the same: servers send/write heartbeats, servers read/consume received heartbeats. The difference is whether they read them from memory or from a topic. However,  heartbeats are transient in nature, a server process them and deletes them. There is no need for persisting them. Moreover, we want to be able to send heartbeats every 100 ms and process heartbeats every 200 ms. With batching and persistence, I don't think a Kafka topic can support such latencies, from the consumer side at least.

The idea with the table is cool but I think there will be too much lag.

Reply:
I think that makes sense - but let's not discredit Kafka here 😂 LinkedIn did a benchmark that showed that the average end-to-end latency on kafka is 2ms and the p99 was 3ms so I think the range of 100 or 200ms is an eternity for a kafka heartbeat to be delivered. source: https://engineering.linkedin.com/kafka/benchmarking-apache-kafka-2-million-writes-second-three-cheap-machines